bundles/org.openhab.binding.irobot/src/main/java/org/openhab/binding/irobot/internal/IdentProtocol.java

```diff
@@ -0,0 +1,131 @@
+ */
+package org.openhab.binding.irobot.internal;
+
+import java.io.IOException;
+import java.io.StringReader;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.InetAddress;
+import java.nio.charset.StandardCharsets;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonParseException;
+import com.google.gson.stream.JsonReader;
+
+/**
+ * iRobot discovery and identification protocol
+ *
+ * @author Pavel Fedin <pavel_fedin@mail.ru> - Initial contribution
+ *
+ */
+public class IdentProtocol {
+
+    private static final String UDP_PACKET_CONTENTS = "irobotmcs";
+    private static final int REMOTE_UDP_PORT = 5678;
+
+    public static DatagramSocket sendRequest(InetAddress host) throws IOException {
+        DatagramSocket socket = new DatagramSocket();
+
+        socket.setBroadcast(true);
+        socket.setReuseAddress(true);
+
+        byte[] packetContents = UDP_PACKET_CONTENTS.getBytes(StandardCharsets.UTF_8);
+        DatagramPacket packet = new DatagramPacket(packetContents, packetContents.length, host, REMOTE_UDP_PORT);
+
+        socket.send(packet);
+        return socket;
+    }
+
+    public static DatagramPacket receiveResponse(DatagramSocket socket) throws IOException {
+        byte[] buffer = new byte[1024];
+        DatagramPacket incomingPacket = new DatagramPacket(buffer, buffer.length);
+
+        socket.setSoTimeout(1000 /* one second */);
+        socket.receive(incomingPacket);
+
+        return incomingPacket;
+    }
+
+    public static IdentData decodeResponse(DatagramPacket packet) throws JsonParseException, ClassCastException {
+        /*
+         * packet is a JSON of the following contents (addresses are undisclosed):
+         * @formatter:off
+         * {
+         *   "ver":"3",
+         *   "hostname":"Roomba-3168820480607740",
+         *   "robotname":"Roomba",
+         *   "ip":"XXX.XXX.XXX.XXX",
+         *   "mac":"XX:XX:XX:XX:XX:XX",
+         *   "sw":"v2.4.6-3",
+         *   "sku":"R981040",
+         *   "nc":0,
+         *   "proto":"mqtt",
+         *   "cap":{
+         *     "pose":1,
+         *     "ota":2,
+         *     "multiPass":2,
+         *     "carpetBoost":1,
+         *     "pp":1,
+         *     "binFullDetect":1,
+         *     "langOta":1,
+         *     "maps":1,
+         *     "edge":1,
+         *     "eco":1,
+         *     "svcConf":1
+         *   }
+         * }
+         * @formatter:on
+         */
+        String reply = new String(packet.getData());
+        // We are not consuming all the fields, so we have to create the reader explicitly
+        // If we use fromJson(String) or fromJson(java.util.reader), it will throw
+        // "JSON not fully consumed" exception, because not all the reader's content has been
+        // used up. We want to avoid that for compatibility reasons because newer iRobot versions
+        // may add fields.
+        JsonReader jsonReader = new JsonReader(new StringReader(reply));
```

Comment:
Can you specify the encoding you expect? E.g. StandardCharsets.UTF_8

Reply:
Done